dali/pipeline/data/backend.cc

```diff
@@ -14,164 +14,161 @@
 
 #include "dali/pipeline/data/backend.h"
 
+#include <atomic>
 #include <mutex>
 #include <memory>
 #include <unordered_map>
 #include <utility>
+#include <vector>
 
 #include "dali/core/common.h"
 #include "dali/pipeline/data/allocator.h"
 #include "dali/pipeline/operator/op_spec.h"
 
 namespace dali {
 
-#define MAP_INSERT_OR_UPDATE(VALUE)                     \
-  int dev;                                              \
-  CUDA_CALL(cudaGetDevice(&dev));                       \
-  auto it = gpu_allocators_.find(dev);                  \
-  if (it != gpu_allocators_.end()) {                    \
-    it->second = VALUE;                                 \
-  } else {                                              \
-    gpu_allocators_.insert(std::make_pair(dev, VALUE)); \
-  }
-
 class AllocatorManager {
  public:
-  static void SetAllocators(const OpSpec &cpu_allocator,
+  AllocatorManager(): cpu_allocator_{nullptr}, pinned_cpu_allocator_{nullptr} {
+    int dev_cout = 0;
+    CUDA_CALL(cudaGetDeviceCount(&dev_cout));
+    gpu_allocators_ = std::vector<std::atomic<GPUAllocator*>>(dev_cout);
+    for (auto& gpu_alloc : gpu_allocators_) {
+      gpu_alloc.store(nullptr);
+    }
+  }
+
+  ~AllocatorManager() {
+    delete cpu_allocator_;
```

Comment:
On one hand - I know it's a destructor and nothing should follow it - but on the other, just for debugging purposes I prefer to set the deleted pointers to nullptr. That should save us some headache when some crazy bug hits us with a rogue thread trying to use these after we've shut down.

Reply:
done

Reply:
```suggestion
    delete std::atomic_exchange(&cpu_allocator_, nullptr);
```